README.md
# The Open Model Project ![](https://img.shields.io/badge/Keep%20a%20Changelog-v1.0.0-brightgreen.svg?logo=data%3Aimage%2Fsvg%2Bxml%3Bbase64%2CPHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGZpbGw9IiNmMTVkMzAiIHZpZXdCb3g9IjAgMCAxODcgMTg1Ij48cGF0aCBkPSJNNjIgN2MtMTUgMy0yOCAxMC0zNyAyMmExMjIgMTIyIDAgMDAtMTggOTEgNzQgNzQgMCAwMDE2IDM4YzYgOSAxNCAxNSAyNCAxOGE4OSA4OSAwIDAwMjQgNCA0NSA0NSAwIDAwNiAwbDMtMSAxMy0xYTE1OCAxNTggMCAwMDU1LTE3IDYzIDYzIDAgMDAzNS01MiAzNCAzNCAwIDAwLTEtNWMtMy0xOC05LTMzLTE5LTQ3LTEyLTE3LTI0LTI4LTM4LTM3QTg1IDg1IDAgMDA2MiA3em0zMCA4YzIwIDQgMzggMTQgNTMgMzEgMTcgMTggMjYgMzcgMjkgNTh2MTJjLTMgMTctMTMgMzAtMjggMzhhMTU1IDE1NSAwIDAxLTUzIDE2bC0xMyAyaC0xYTUxIDUxIDAgMDEtMTItMWwtMTctMmMtMTMtNC0yMy0xMi0yOS0yNy01LTEyLTgtMjQtOC0zOWExMzMgMTMzIDAgMDE4LTUwYzUtMTMgMTEtMjYgMjYtMzMgMTQtNyAyOS05IDQ1LTV6TTQwIDQ1YTk0IDk0IDAgMDAtMTcgNTQgNzUgNzUgMCAwMDYgMzJjOCAxOSAyMiAzMSA0MiAzMiAyMSAyIDQxLTIgNjAtMTRhNjAgNjAgMCAwMDIxLTE5IDUzIDUzIDAgMDA5LTI5YzAtMTYtOC0zMy0yMy01MWE0NyA0NyAwIDAwLTUtNWMtMjMtMjAtNDUtMjYtNjctMTgtMTIgNC0yMCA5LTI2IDE4em0xMDggNzZhNTAgNTAgMCAwMS0yMSAyMmMtMTcgOS0zMiAxMy00OCAxMy0xMSAwLTIxLTMtMzAtOS01LTMtOS05LTEzLTE2YTgxIDgxIDAgMDEtNi0zMiA5NCA5NCAwIDAxOC0zNSA5MCA5MCAwIDAxNi0xMmwxLTJjNS05IDEzLTEzIDIzLTE2IDE2LTUgMzItMyA1MCA5IDEzIDggMjMgMjAgMzAgMzYgNyAxNSA3IDI5IDAgNDJ6bS00My03M2MtMTctOC0zMy02LTQ2IDUtMTAgOC0xNiAyMC0xOSAzN2E1NCA1NCAwIDAwNSAzNGM3IDE1IDIwIDIzIDM3IDIyIDIyLTEgMzgtOSA0OC0yNGE0MSA0MSAwIDAwOC0yNCA0MyA0MyAwIDAwLTEtMTJjLTYtMTgtMTYtMzEtMzItMzh6bS0yMyA5MWgtMWMtNyAwLTE0LTItMjEtN2EyNyAyNyAwIDAxLTEwLTEzIDU3IDU3IDAgMDEtNC0yMCA2MyA2MyAwIDAxNi0yNWM1LTEyIDEyLTE5IDI0LTIxIDktMyAxOC0yIDI3IDIgMTQgNiAyMyAxOCAyNyAzM3MtMiAzMS0xNiA0MGMtMTEgOC0yMSAxMS0zMiAxMXptMS0zNHYxNGgtOFY2OGg4djI4bDEwLTEwaDExbC0xNCAxNSAxNyAxOEg5NnoiLz48L3N2Zz4K) [![Netlify Status](https://api.netlify.com/api/v1/badges/abb386d0-844f-4751-9647-dc544a7b6360/deploy-status)](https://app.netlify.com/sites/fervent-albattani-780d1a/deploys)


We believe that better election forecasting and transparency are important. Play with the inputs to our model to understand how it works,
[download for more options](https://docs.google.com/spreadsheets/d/1BM-HSoegMc6gxu56LCNnpIlm_b6jlDEtztiQ4EF5jy4/copy), and [join our discord server](https://discord.gg/HrnsJZ5baM) to discuss.

![The Open Model Project](https://github.com/ExtraE113/OpenModelProject/blob/master/images/Screenshot_2020-12-21%20The%20Open%20Model%20Project.png)

## Installation
This website is built with Jekyll.

### Installing Ruby & Jekyll
 
If this is your first time using Jekyll, please follow the [Jekyll docs](https://jekyllrb.com/docs/installation/) and make sure your local environment (including Ruby) is setup correctly.

### Installing Repo

Download or clone the repository.

To run the website locally, navigate to the theme directory and run:

```
bundle install
``` 

To start the Jekyll local development server.

```
bundle exec jekyll serve
``` 

To build the site.
 
```
bundle exec jekyll build
```

## Credits

- Beautiful royalty free Illustrations by [Icons8](https://icons8.com/illustrations/style--pixeltrue) and [undraw.co](https://undraw.co)
- Theme by [Robert Austin](https://jamstackthemes.dev/theme/jekyll-serif/)


CHANGELOG.md
---
title: 'Changelog'
layout: page
---


All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.8.1] - 2021-1-06
### Changed
Home page updated  

## [1.8.0] - 2021-1-06
### Added
Added preliminary postmortem  
### Changed
Home page now links to that ^^^  
Model now shows actual MOV  

## [1.7.3] - 2021-1-05
### Removed
Removed banner & live section

## [1.7.2] - 2021-1-05
### Added
Added banner with a link to live updates

## [1.7.1] - 2021-1-05
### Changed
Decreased top padding  
Added pod buttons

## [1.7.0] - 2021-1-05
### Changed
Copyright year updated  
Added podcast info

## [1.6.1] - 2020-12-31
### Added
Added last updated (in JS, because that's where we go to update the stats)

## [1.6.0] - 2020-12-31
### Added
Add option to weight by exit poll data (2020) instead of by census


## [1.5.4] - 2020-12-28
### Changed
To prevent div by 0 errors on the backend set the minimum slider value to 0.01

## [1.5.3] - 2020-12-28
### Changed
header and footer

## [1.5.2] - 2020-12-28
### Changed
updated memeo  

## [1.5.1] - 2020-12-28
### Changed
experimental factors disabled by default


## [1.5.0] - 2020-12-28
### Changed
/polling-method is now /ga-poll-runoff-method

## [1.4.9] - 2020-12-28
### Changed
copy fixes  

## [1.4.8] - 2020-12-28
### Changed
fix form problem with netlify  
added image for tad  
changed image for ezra  

## [1.4.7] - 2020-12-28
### Changed
Replaced call box with contact form  
Changed copyright text  
Added twitter info  
Added patreon info  

## [1.4.6] - 2020-12-28
### Changed
Formatting tweak on features page

## [1.4.5] - 2020-12-28
### Changed
Many copy fixes

## [1.4.4] - 2020-12-28
### Changed
Changed language RE: polling date and sample size

## [1.4.3] - 2020-12-28
### Changed  
Changed methodology sections

## [1.4.2] - 2020-12-27
### Added
Fixed typo that caused warnock always to have a 0% chance  
Added some links

## [1.4.1] - 2020-12-27
### Added
MOV shows up
Latest model changes
Minor tweaks to get everything playing nice


## [1.4.0] - 2020-12-26
### Added  
preview subdomain 

## [1.3.4] - 2020-12-27
### Changed
Uses production API endpoint

## [1.3.3] - 2020-12-27
### Changed  
Changed previous update date in CHANGELOG to reflect reality  
Minor language tweak for the inhouse poll

## [1.3.2] - 2020-12-26
### Changed
Minor update to use much faster api

## [1.3.1] - 2020-12-24
### Changed
Language in methodology section is now correct

## [1.3.0] - 2020-12-24
### Added
poll explorer, based off API that re-executes code every time  
Removed changelog from ignore and tweaked formatting

## [1.2.0] - 2020-12-24
### Changed
On-website options changed
`model-switcher.js` improved

## [1.1.5] - 2020-12-23
### Added
Twitter SEO support @OpenModelProj

## [1.1.4] - 2020-12-22
### Added
404 page

## [1.1.3] - 2020-12-22
### Changed
Change handler now called on dom ready so #s are only in one place

## [1.1.2] - 2020-12-22
### Changed
Removed inconsistent question mark

## [1.1.1] - 2020-12-22
### Changed
VIEW ALL PREDICTIONS CHANGE ALL INPUTS now breaks on small devices


## [1.1.0] - 2020-12-22
### Added
Tweak the model assumptions now works
### Changed
Hosting now on netlify

## [1.0.2] - 2020-12-21
### Changed
Intro image looks a lot better on varying screen sizes.
Tweaked padding for static predictions.

## [1.0.1] - 2020-12-21
### Changed
Harry Crane's image is now served from our website, so it's served over https

## [1.0.0] - 2020-12-21
### Added
Static odds readout

### Changed
Fixed a bug where depending on build conditions the order of the features could change
Changed email and removed phone number from contact
Added in and fixed some broken SEO stuff.


assets/js/points.js
let surveyResponses = [
    ["", 0, 0, 0, 0, 1,  //page 1
        3, 1, //page 2
        1, //page 3
        null, //page 4
        0, //page 5
        1, 1, //page 6
        null, null, //page 7
        null, //page 8
        0, 1, 0,//page 9
        0, 3, 4, 5, //page 10
        1, 1, 1, 1, //page 11
        "FTI", // page 12
        0, 2000, 0, 12, 2, //page 13
        0, 0, 0, 0, 0, //page 14
        1, //page 15
        0 //points
    ]
]

function calculateLikelyhoodPoints() {
    for (const response of surveyResponses) {
        let shortTexts = getShortTexts(response);

        console.log(shortTexts);

        for (const shortText of shortTexts) {

            if(shortText === "alreadyVoted" && document.getElementById("alreadyVoted").checked) {
                response[response.length - 1] = -1; //todo get max and set all elements with -1 to max
                break;
            }

            if(shortText === "notRegistered" && document.getElementById("notRegistered").checked) {
                response[response.length - 1] = 0;
                break;
            }

            if(shortText === "nothonest" || shortText === "somewhathonest") {
                continue;
                //todo
            }
            response[response.length-1] += parseFloat(document.getElementById(shortText).value);
        }

    }

    console.log(surveyResponses);
}

//gets the short id texts (used as the input's id=) given a single response object (arr of responses)
function getShortTexts(response) {
    let out = []

    const texts = [
        null, ["thinkingALot", "thinkingSome", "thinkingLittle"], null, [null, "notRegistered"], null, ["alreadyVoted", null, null], //page 1
        ["voteInPersonOrRecieved", "voteInPersonOrRecieved", "voteInPersonOrRecieved", "requestedNotReceived", "notRequested", null, null], ["veryLikely", "likely", "neitherLikelyNorUnlikely", "somewhatUnlikely", "unlikely"], //page 2
        null, //page 3
        null, //page 4
        null, //page 5
        null, null, //page 6
        null, null, //page 7
        null, //page 8
        null, null, null,//page 9
        null, null, null, null, //page 10
        null, null, null, null, //page 11
        (function (res) { return res.includes("Y") ? "excludeIncorrect" : null}), // page 12
        null, null, null, null, null, //page 13
        null, null, null, null, null, //page 14
        ["nothonest", "somewhathonest", null, null] //page 15
    ];


    for (let i = 0; i < texts.length; i++) {
        if(texts[i] === null){
            continue
        }
        if(Array.isArray(texts[i])){
            if (texts[i][response[i]] !== null)
                out.push(texts[i][response[i]])
            continue
        }
        if(typeof texts[i] === "function"){
            let result = texts[i](response[i])
            if (result !== null)
                out.push(result);
        }
    }

    return out;

}

assets/js/checkbox.js
function registerCheckboxUpdateHandler() {
    const checkboxGroups = [
        {
            "id": "census-overall",
            "class": "census-data",
        },
        {
            "id": "past-vote-overall",
            "class": "past-vote-data",
        },
        {
            "id": "experimental-overall",
            "class": "experimental-data",
        }
    ]

    for (const cbGroup of checkboxGroups) {
        let checkboxes = document.querySelectorAll('input.subOption.' + cbGroup["class"]),
            checkall = document.getElementById(cbGroup["id"]);

        for (let i = 0; i < checkboxes.length; i++) {
            checkboxes[i].onclick = function () {
                let checkedCount = document.querySelectorAll('input.subOption.' + cbGroup["class"] + ':checked').length;

                checkall.checked = checkedCount > 0;
                checkall.indeterminate = checkedCount > 0 && checkedCount < checkboxes.length;
            }
        }

        checkall.onclick = function () {
            for (let i = 0; i < checkboxes.length; i++) {
                checkboxes[i].checked = this.checked;
            }
        }
    }
}


assets/js/poll-switcher.js
function handleChange() {
    let traditionalWeight = document.getElementById("weighting").value.includes("Traditional")
    let likelyModel = document.getElementById("likely").value.includes("Likely")

    let scenario = (traditionalWeight ? "t" : "f") + (likelyModel ? "l" : "r")

    //senario: [ossoff margin, ossoff MOE, warnock margin, warnock MOE]
    let data = {
        "tr": [1.3, 4.6, 2.0, 4.6],
        "tl": [-3.8, 4.7, -3.0, 4.7],
        "fr": [3.6, 4.6, 4.1, 4.6],
        "fl": [-1.9, 4.7, -1.4, 4.7],
    }

    let [ossoff, ossoffMOE, warnock, warnockMOE] = data[scenario];


    document.getElementById("special-headline").innerHTML = (warnock >= 0 ? "Warnock +" : "Loeffler +") + (Math.round(Math.abs(warnock) * 100) / 100).toFixed(1) + "%";
    document.getElementById("special-margin").innerHTML = "&#177;" + warnockMOE;
    document.getElementById("regular-headline").innerHTML = (ossoff >= 0 ? "Ossoff +" : "Perdue +") + (Math.round(Math.abs(ossoff) * 100) / 100).toFixed(1) + "%";
    document.getElementById("regular-margin").innerHTML = "&#177;" + ossoffMOE;


}

function ready() {
    handleChange();
}

document.addEventListener("DOMContentLoaded", ready, false)

assets/js/model-switcher.js
function handleChange() {
    let covid = document.getElementById("covid").checked
    let market = document.getElementById("market").checked
    let prior = document.getElementById("prior").checked
    let avg = document.getElementById("average").value

    let scenario = (covid ? "c" : "x") + (avg.includes("imple") ? "s" : "w") + (market ? "p" : "x") + (prior ? "e" : "x")

    let data = {
        "xsxx": [0.51, 0.53, "Ossoff margin: 0.3 (80% CI: -3.6 to 4.1)", 0.62, "Warnock margin: 1 (80% CI: -2.9 to 4.8)"],
        "csxx": [0.37, 0.39, "Ossoff margin: -0.9 (80% CI: -4.8 to 3)", 0.48, "Warnock margin: -0.2 (80% CI: -4.1 to 3.7)"],
        "xwxx": [0.66, 0.71, "Ossoff margin: 1.8 (80% CI: -2.1 to 5.6)", 0.71, "Warnock margin: 1.8 (80% CI: -2 to 5.7)"],
        "cwxx": [0.49, 0.54, "Ossoff margin: 0.4 (80% CI: -3.5 to 4.2)", 0.55, "Warnock margin: 0.4 (80% CI: -3.4 to 4.3)"],
        "xspx": [0.51, 0.53, "Ossoff margin: 0.3 (80% CI: -3.6 to 4.1)", 0.62, "Warnock margin: 1 (80% CI: -2.9 to 4.8)"],
        "cspx": [0.39, 0.4, "Ossoff margin: -0.9 (80% CI: -4.8 to 3)", 0.49, "Warnock margin: -0.2 (80% CI: -4.1 to 3.7)"],
        "xwpx": [0.62, 0.69, "Ossoff margin: 1.8 (80% CI: -2.1 to 5.6)", 0.7, "Warnock margin: 1.8 (80% CI: -2 to 5.7)"],
        "cwpx": [0.48, 0.54, "Ossoff margin: 0.4 (80% CI: -3.5 to 4.2)", 0.56, "Warnock margin: 0.4 (80% CI: -3.4 to 4.3)"],
        "xsxe": [0.39, 0.41, "Ossoff margin: -0.8 (80% CI: -4.6 to 3.1)", 0.5, "Warnock margin: 0 (80% CI: -3.8 to 3.9)"],
        "csxe": [0.33, 0.34, "Ossoff margin: -1.3 (80% CI: -5.2 to 2.5)", 0.43, "Warnock margin: -0.5 (80% CI: -4.4 to 3.3)"],
        "xwxe": [0.47, 0.5, "Ossoff margin: 0 (80% CI: -3.9 to 3.8)", 0.56, "Warnock margin: 0.5 (80% CI: -3.4 to 4.3)"],
        "cwxe": [0.38, 0.41, "Ossoff margin: -0.7 (80% CI: -4.6 to 3.1)", 0.47, "Warnock margin: -0.2 (80% CI: -4.1 to 3.6)"],
        "xspe": [0.41, 0.42, "Ossoff margin: -0.8 (80% CI: -4.6 to 3.1)", 0.52, "Warnock margin: 0 (80% CI: -3.8 to 3.9)"],
        "cspe": [0.35, 0.36, "Ossoff margin: -1.3 (80% CI: -5.2 to 2.5)", 0.45, "Warnock margin: -0.5 (80% CI: -4.4 to 3.3)"],
        "xwpe": [0.47, 0.5, "Ossoff margin: 0 (80% CI: -3.9 to 3.8)", 0.56, "Warnock margin: 0.5 (80% CI: -3.4 to 4.3)"],
        "cwpe": [0.4, 0.42, "Ossoff margin: -0.7 (80% CI: -4.6 to 3.1)", 0.48, "Warnock margin: -0.2 (80% CI: -4.1 to 3.6)"]
    }

    let last_updated = "2021-01-05 10AM ET"

    let [dsenate, ossoff, ossoff_margin, warnock, warnock_margin] = data[scenario];


    document.getElementById("dsenate").innerText = Math.round(dsenate * 100) + "%";
    document.getElementById("ossoff").innerText = Math.round(ossoff * 100) + "%";
    document.getElementById("warnock").innerText = Math.round(warnock * 100) + "%";

    document.getElementById("ossoff-mov").innerText = ossoff_margin;
    document.getElementById("warnock-mov").innerText = warnock_margin;

    document.getElementById("last-updated").innerText = "Last updated " + last_updated

}

function ready() {
    handleChange();
    let elements = document.getElementsByClassName("model-switcher")
    for (const element of elements) {
        element.addEventListener("change", handleChange)
    }
}

document.addEventListener("DOMContentLoaded", ready, false)


assets/js/prefomAllCalculations.js
var outputsUpdated = true;
var updateIndexRequest = 0;
var old_body = null;
var old_response = null;

function updateOutputs() {
    outputsUpdated = false;
    // calculateLikelyhoodPoints();
    updateSliderFeedback();

    outputsUpdated = true;
}


function getUpdatedValues() {


    // ensure that updateOutputs runs before this (to ensure checkboxes work). This is a horrible, hacky way to do this
    // and needs to be rewritten but it works for now
    /*
       ^(;,;)^
     */
    if (outputsUpdated === false) {
        window.setTimeout(getUpdatedValues, 100); /* this checks the flag every 100 milliseconds*/
    } else {

        outputsUpdated = false;

        document.getElementById("regular-headline").innerText = ""
        document.getElementById("regular-margin").innerText = ""
        document.getElementById("special-headline").innerText = ""
        document.getElementById("special-margin").innerText = ""

        for (const item of document.querySelectorAll(".ph-item")) {
            item.classList.remove("d-none")
        }


        let weigh_on = Array.from(document.querySelectorAll('input.subOption.weight:checked')).map(
            function (element) {
                return element.id;
            }
        )

        const weigh_on_key = ['gender', 'race', 'education', 'income', 'age', 'vote2016', 'vote2020', 'loc_county',
            'gss_trust', 'gss_bible', 'gss_spanking', 'social_fb'];

        let weigh_on_code = ""

        for (let i = 0; i < weigh_on_key.length; i++) {
            if (weigh_on.includes(weigh_on_key[i])) {
                weigh_on_code += "1"
            } else {
                weigh_on_code += "0"
            }
        }


        let body = {
            "very likely": parseFloat(document.getElementById("veryLikely").value),
            "already voted": parseFloat(document.getElementById("alreadyVoted").value),
            "likely": parseFloat(document.getElementById("likely").value),
            "somewhat likely": parseFloat(document.getElementById("somewhatLikely").value),
            "neither likely": parseFloat(document.getElementById("neitherLikelyNorUnlikely").value),
            "somewhat unlikely": parseFloat(document.getElementById("somewhatUnlikely").value),
            "unlikely": parseFloat(document.getElementById("unlikely").value),
            "2020 voted": parseFloat(document.getElementById("turnout2020").value),
            "weigh_on_code": weigh_on_code,
            "exits": document.getElementById("exits").checked ? "True" : "False"
        }


        if (JSON.stringify(old_body) === JSON.stringify(body)) {
            console.log("here")
            handleResponse(old_response)
        } else {
            old_body = body;
            postData("https://udshih2lj8.execute-api.us-east-1.amazonaws.com/prod5/inhousepolls/weight",
                body).then(handleResponse).catch()
        }


    }
}

function handleResponse(data) {
    old_response = data;
    for (const item of document.querySelectorAll(".ph-item")) {
        item.classList.add("d-none")
    }

    let regular = document.getElementById("likely-voter-on").checked ? data["ossoff_perdue_demo_likely"] : data["ossoff_perdue_demo"]
    document.getElementById("regular-headline").innerText = (regular["headline"] > 0 ? "Ossoff +" : "Perdue +") + (Math.round(Math.abs(regular["headline"]) * 100) / 100).toFixed(1) + "%"
    document.getElementById("regular-margin").innerHTML = "&#177;" + (Math.round(Math.abs(regular["moe"]) * 100) / 100).toFixed(1)

    let special = document.getElementById("likely-voter-on").checked ? data["warnock_loeffler_demo_likely"] : data["warnock_loeffler_demo"]

    document.getElementById("special-headline").innerText = (special["headline"] > 0 ? "Warnock +" : "Loeffler +") + (Math.round(Math.abs(special["headline"]) * 100) / 100).toFixed(1) + "%"
    document.getElementById("special-margin").innerHTML = "&#177;" + (Math.round(Math.abs(special["moe"]) * 100) / 100).toFixed(1)
}

function ready() {
    registerCheckboxUpdateHandler();
    getUpdatedValues();
}

document.addEventListener("DOMContentLoaded", ready, false)


// from https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
// Example POST method implementation:
async function postData(url = '', data = {}) {
    updateIndexRequest++;
    let initUpdateIndexRequest = updateIndexRequest;

    // Default options are marked with *
    const response = await fetch(url, {
        method: 'POST', // *GET, POST, PUT, DELETE, etc.
        mode: 'cors', // no-cors, *cors, same-origin
        cache: 'no-cache', // *default, no-cache, reload, force-cache, only-if-cached
        credentials: 'same-origin', // include, *same-origin, omit
        headers: {
            'Content-Type': 'application/json'
            // 'Content-Type': 'application/x-www-form-urlencoded',
        },
        redirect: 'follow', // manual, *follow, error
        referrerPolicy: 'no-referrer', // no-referrer, *no-referrer-when-downgrade, origin, origin-when-cross-origin, same-origin, strict-origin, strict-origin-when-cross-origin, unsafe-url
        body: JSON.stringify(data) // body data type must match "Content-Type" header
    });

    //this function has been called again so we can discard. todo https://stackoverflow.com/questions/31061838/how-do-i-cancel-an-http-fetch-request
    if (updateIndexRequest !== initUpdateIndexRequest) {
        throw Error("no longer needed")
    }

    return response.json(); // parses JSON response into native JavaScript objects
}

assets/js/model-switcher-archive.js
function handleChange() {
    let covid = document.getElementById("covid").checked
    let market = document.getElementById("market").checked
    let prior = document.getElementById("prior").checked
    let avg = document.getElementById("average").value

    let scenario = (covid ? "c" : "x") + (avg.includes("imple") ? "s" : "w") + (market ? "p" : "x") + (prior ? "e" : "x")

    let data = {
        "xsxx": [0.51, 0.53, "Margin: 0.3 (80% CI: -3.6 to 4.1)", 0.62, "Margin: 1 (80% CI: -2.9 to 4.8)"],
        "csxx": [0.37, 0.39, "Margin: -0.9 (80% CI: -4.8 to 3)", 0.48, "Margin: -0.2 (80% CI: -4.1 to 3.7)"],
        "xwxx": [0.66, 0.71, "Margin: 1.8 (80% CI: -2.1 to 5.6)", 0.71, "Margin: 1.8 (80% CI: -2 to 5.7)"],
        "cwxx": [0.49, 0.54, "Margin: 0.4 (80% CI: -3.5 to 4.2)", 0.55, "Margin: 0.4 (80% CI: -3.4 to 4.3)"],
        "xspx": [0.51, 0.53, "Margin: 0.3 (80% CI: -3.6 to 4.1)", 0.62, "Margin: 1 (80% CI: -2.9 to 4.8)"],
        "cspx": [0.39, 0.4, "Margin: -0.9 (80% CI: -4.8 to 3)", 0.49, "Margin: -0.2 (80% CI: -4.1 to 3.7)"],
        "xwpx": [0.62, 0.69, "Margin: 1.8 (80% CI: -2.1 to 5.6)", 0.7, "Margin: 1.8 (80% CI: -2 to 5.7)"],
        "cwpx": [0.48, 0.54, "Margin: 0.4 (80% CI: -3.5 to 4.2)", 0.56, "Margin: 0.4 (80% CI: -3.4 to 4.3)"],
        "xsxe": [0.39, 0.41, "Margin: -0.8 (80% CI: -4.6 to 3.1)", 0.5, "Margin: 0 (80% CI: -3.8 to 3.9)"],
        "csxe": [0.33, 0.34, "Margin: -1.3 (80% CI: -5.2 to 2.5)", 0.43, "Margin: -0.5 (80% CI: -4.4 to 3.3)"],
        "xwxe": [0.47, 0.5, "Margin: 0 (80% CI: -3.9 to 3.8)", 0.56, "Margin: 0.5 (80% CI: -3.4 to 4.3)"],
        "cwxe": [0.38, 0.41, "Margin: -0.7 (80% CI: -4.6 to 3.1)", 0.47, "Margin: -0.2 (80% CI: -4.1 to 3.6)"],
        "xspe": [0.41, 0.42, "Margin: -0.8 (80% CI: -4.6 to 3.1)", 0.52, "Margin: 0 (80% CI: -3.8 to 3.9)"],
        "cspe": [0.35, 0.36, "Margin: -1.3 (80% CI: -5.2 to 2.5)", 0.45, "Margin: -0.5 (80% CI: -4.4 to 3.3)"],
        "xwpe": [0.47, 0.5, "Margin: 0 (80% CI: -3.9 to 3.8)", 0.56, "Margin: 0.5 (80% CI: -3.4 to 4.3)"],
        "cwpe": [0.4, 0.42, "Margin: -0.7 (80% CI: -4.6 to 3.1)", 0.48, "Margin: -0.2 (80% CI: -4.1 to 3.6)"]
    }

    let last_updated = "2021-01-05 10AM ET"
    let last_updated_actual = "2021-01-08 1:07 PM ET"

    let ossoff_margin_actual = "+1.1"
    let warnock_margin_actual = "+2.0"

    let [dsenate, ossoff, ossoff_margin, warnock, warnock_margin] = data[scenario];

    document.getElementById("dsenate").innerText = Math.round(dsenate * 100) + "%";
    document.getElementById("ossoff").innerText = Math.round(ossoff * 100) + "%";
    document.getElementById("warnock").innerText = Math.round(warnock * 100) + "%";

    document.getElementById("ossoff-mov").innerText = ossoff_margin;
    document.getElementById("warnock-mov").innerText = warnock_margin;

    document.getElementById("ossoff-mov-actual").innerText = "Actual MOV " + ossoff_margin_actual;
    document.getElementById("warnock-mov-actual").innerText = "Actual MOV " + warnock_margin_actual;

    document.getElementById("last-updated").innerText = "Last updated " + last_updated
    document.getElementById("last-updated-actual").innerText = "Actual vote numbers last updated " + last_updated_actual + "."

}

function ready() {
    handleChange();
    let elements = document.getElementsByClassName("model-switcher")
    for (const element of elements) {
        element.addEventListener("change", handleChange)
    }
}

document.addEventListener("DOMContentLoaded", ready, false)
